src/server/dflycmd.cc

```diff
@@ -380,79 +372,105 @@ void DflyCmd::FullSyncFb(FlowInfo* flow) {
   }
 
   if (ec) {
-    LOG(ERROR) << ec;
-    return;
+    return cntx->Error(ec);
   }
 
-  // TODO: we should be able to stop earlier if requested.
-  ec = saver->SaveBody(nullptr);
-  if (ec) {
-    LOG(ERROR) << ec;
-    return;
+  if ((ec = saver->SaveBody(*cntx, nullptr))) {
+    return cntx->Error(ec);
   }
 
-  VLOG(1) << "Sending full sync EOF";
-
   ec = flow->conn->socket()->Write(io::Buffer(flow->eof_token));
   if (ec) {
-    LOG(ERROR) << ec;
-    return;
+    return cntx->Error(ec);
   }
 }
 
 uint32_t DflyCmd::CreateSyncSession() {
   unique_lock lk(mu_);
+  unsigned sync_id = next_sync_id_++;
 
-  auto sync_info = make_shared<SyncInfo>();
-  sync_info->flows.resize(shard_set->size() + 1);
+  unsigned flow_count = shard_set->size() + 1;
+  auto err_handler = [this, sync_id](const GenericError& err) {
+    LOG(INFO) << "Running error handler for sync session " << sync_id;
 
-  auto [it, inserted] = sync_infos_.emplace(next_sync_id_, std::move(sync_info));
+    // Stop replication in case of error.
+    // StopReplication needs to run async to prevent blocking
+    // the error handler.
+    ::boost::fibers::fiber{&DflyCmd::StopReplication, this, sync_id}.detach();
+
+    return true;  // Cancel context
+  };
+
+  auto sync_info = make_shared<SyncInfo>(flow_count, std::move(err_handler));
+  auto [it, inserted] = sync_infos_.emplace(sync_id, std::move(sync_info));
   CHECK(inserted);
 
-  return next_sync_id_++;
+  return sync_id;
 }
 
-void DflyCmd::UnregisterFlow(FlowInfo* flow) {
-  // TODO: Cancel saver operations.
-  flow->conn = nullptr;
-  flow->saver.reset();
-}
+void DflyCmd::OnClose(ConnectionContext* cntx) {
+  unsigned session_id = cntx->conn_state.repl_session_id;
+  if (!session_id)
+    return;
 
-void DflyCmd::DeleteSyncSession(uint32_t sync_id) {
-  shared_ptr<SyncInfo> sync_info;
+  auto sync_info = GetSyncInfo(session_id);
+  if (!sync_info)
+    return;
 
-  // Remove sync_info from map.
-  // Store by value to keep alive.
-  {
-    unique_lock lk(mu_);
+  // Because cancellation holds the mutex,
+  // this lock waits for it to finish before releasing the underlying connection
+  lock_guard lk(sync_info->mu);
```

Comment:
I would move mutex locking into CancelSyncSession - you test the state there anyway.